File: Excel/563 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/563 Bands of Numbers.xlsx"
input = read_excel(path, range = "A2:B17")
test  = read_excel(path, range = "D2:F6")

result = input %>%
  mutate(Group = cumsum(c(1, diff(Numbers)) != 1), .by = Product) %>%
  mutate(Band = ifelse(n() == 1, paste0(Numbers), paste0(Numbers[1], "-", Numbers[n()])), 
         .by = c(Product, Group)) %>%
  summarise(Bands = paste0(unique(Band), collapse = ", "), 
            Count = n_distinct(Band), 
            .by = Product)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE

File: Excel/500-599/563/563 Challenge.py
import pandas as pd

path = "563 Bands of Numbers.xlsx"
input = pd.read_excel(path, usecols="A:B", skiprows=1, nrows=16)
test = pd.read_excel(path, usecols="D:F", skiprows=1, nrows=4)\
    .rename(columns=lambda x: x.replace('.1', ''))

result = input.copy()
result['Group'] = result.groupby('Product')['Numbers'].diff().ne(1).cumsum()
result['Bands'] = result.groupby(['Group', 'Product'])['Numbers']\
    .transform(lambda x: f"{x.min()}-{x.max()}" if x.count() > 1 else x).astype(str)

result = result.groupby('Product').agg(
    Band=('Bands', lambda x: ', '.join(x.unique())),
    Count=('Bands', 'nunique')
).reset_index()

print(result.equals(test))  # True
